### PR TITLE
Websocket Library

### PR DESCRIPTION
**Pull request by pixelblend**
This PR is an example of what a shared Websocket Library could look like.

Functionality:
- Can subscribe to a single topic, or to a regex group of topics
- Publish to a single topic
- Validates topics on subscribe & publish
- Handles JSON marshalling

TODO:
- [x] Mount as an app
- [x] Add to remaining apps
- [x] Make available to services as a lib
- [x] Add to all services

### CODE REVIEW

File: apps/websocket/external [hunk 1]
```diff
@@ -0,0 +1 @@
+./internal
```
Comment (pixelblend) [inline apps/websocket/external]: This is a symlink so I don't have to duplicate code for both sides of the apps

File: apps/websocket/internal/index.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  const onOpen = (event) => log('connected', event);
+
+  const onMessage = (event) => {
+    const { topic, payload } = JSON.parse(event.data);
+
+    subscriptions.matching(topic).forEach(cb => cb({ topic, payload }));
+  };
+
+  const subscribe = (topic, cb) => {
+    subscriptions.subscribe(topic, cb);
```
Comment (andrewn) [inline apps/websocket/internal/index.js]: Does this imply clients can only have one callback per topic? If so, if a client calls subscribe again, is it obvious that their last callback will no longer work?

File: apps/radio/internal/index.html [hunk 3]
```diff
@@ -34,35 +33,35 @@ <h1>This is the part of the radio app that runs in the radio itself</h1>
...
-      const msg = JSON.parse(evt.data);
-      console.log('message', msg);
-      console.log('message key', msg.key);
-      if(msg.key == "play"){
-         var audio_file = msg.data.params.file;
+
+    import createWebsocket from '/websocket';
+    const ws = createWebsocket({ debug: true });
+
+    ws.subscribe(new RegExp('radio/command/.*'), ({ topic, payload }) => {
```
[the remark this answers is not in the document]
Comment (pixelblend) [reply apps/radio/internal/index.html]: I didn't want the extra complexity of having to scan a string for something that might be a wildcard. I would probably have to turn them into regexs internally anyway.

This way is at least clear about what's going on.

File: apps/websocket/internal/index.js (hunk 2 above)
Comment (pixelblend) [reply apps/websocket/internal/index.js]: I could either:

- Allow multiple callbacks
- Throw an error when trying to set an existing sub

Which would you prefer? Multiple callbacks might be a bit confusing?

File: apps/radio/external/index.html [hunk 4]
```diff
@@ -2,32 +2,45 @@
...
-  <h1>This is the public part of the radio app</h1>
-  <p>Useful for remote controls etc</p>
-  <button onClick="send('play', { params: { file: 'audio/Birdsong-Collared_Dove.mp3' }})">Play birdsong - collared dove</button>
-  <button onClick="send('play', { params: { file: 'audio/Birdsong-Blue_Tit.mp3' }})">Play birdsong - blue tit</button>
-  <button onClick="send('hls', {})">Play HLS stream</button>
-  <button onClick="send('chord', {})">Generate Web Audio chord</button>
-
-  <script>
+  <script type="module">
+    import createWebsocket from '/websocket';
```
[the remark this answers is not in the document]
Comment (pixelblend) [reply apps/radio/external/index.html]: `web-socket` ❓

**Comment by pixelblend** [comment]
>Do you have an approach in mind for re-using this in the services too as they also need to subscribe and emit to the WS broker.

I'm hoping to make a few tweaks so the code is isomorphic (🤢), mount it somewhere shared and symlink it across the codebases.

**Comment by pixelblend** [comment]
>Is there a reason the ws object doesn't have an EventEmitter-like API? i.e. `ws.on('radio.command.play')`?

I thought the `pub` / `sub` interface made the most sense for this object. How would you model the publishing for an EventEmitter model?

File: apps/radio/external/index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewn) [reply apps/radio/external/index.html]: Actually `crEateWebSOCKet` can be named anything the client wants. So let's stick with `websocket` as it's easier to remember.

As you were, nothing to see here.

File: apps/websocket/internal/index.js (hunk 2 above)
Comment (andrewn) [reply apps/websocket/internal/index.js]: I guess warning for now would make sense. I just feel like it could be the source of some confusing bugs.

**Comment by andrewn** [comment]
>>Is there a reason the ws object doesn't have an EventEmitter-like API? i.e. ws.on('radio.command.play')?

>I thought the pub / sub interface made the most sense for this object. How would you model the publishing for an EventEmitter model?

Thinking about this more I think this is related to having multiple callbacks per subscription. But since I agree that we don't want those then this doesn't matter?

If we _did_ have multiple callbacks then you'd subscribe to your events via `on(topic, cb)` and then publish via `.publish(topic, payload)`. This is how the WebSocket interface is with `ws.addEventListener()` and `ws.send()`?

**Comment by pixelblend** [comment]
@andrewn added uniqueness to subscriptions ☝️

**Pull request merged**